File: Solutions/0309/0309.js
function maxProfit(prices) {
  let sell = 0;
  let hold = Number.NEGATIVE_INFINITY;
  let prev = 0;

  for (const price of prices) {
    const cache = sell;
    sell = Math.max(sell, hold + price);
    hold = Math.max(hold, prev - price);
    prev = cache;
  }

  return sell;
}

File: Solutions/0309/0309.go
func maxProfit(prices []int) int {
	sell := 0
	hold := math.MinInt64
	prev := 0

	for _, price := range prices {
		cache := sell
		sell = max(sell, hold+price)
		hold = max(hold, prev-price)
		prev = cache
	}

	return sell
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}